Llama Index serde and pyfunc

michael-berk opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/michael-berk/mlflow/pull/11674?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11674/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11674
```

</p>
</details>

### Related Issues/PRs


### What changes are proposed in this pull request?

Commits `10` (ready for review)
* Add serialize objects util
* Add tests for serialize objects util
* Add save/load functionality
* Add simple e2e test on save/load
* Add subs for signature inference - not ready for review

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Official flavor support for llama index. 

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

<!-- patch -->

- [ ] Yes
- [ ] No

Commit:
Model stubs

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Use mocks

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Drop unecessary llm file

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Patch settings

Signed-off-by: Michael Berk <michael.berk@databricks.com>

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,273 @@
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+
+    if signature is None and input_example is not None:
+        wrapped_model = _LlamaIndexModelWrapper(index)
```
Does the input / output schema change for the type of engine that is specified?

Commit:
Settings serialization reset

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Settings serde complete

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Add fixtures for tests. Improve coverage of object serde. Prompt template serde

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Working load/log

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Reset for new PR

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
test.py

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
reset

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
10: save and load

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
10: fix tests linter remove unused imports

Signed-off-by: Michael Berk <michael.berk@databricks.com>

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+        mlflow_model.metadata = metadata
+
+    engine_kwargs_path = os.path.join(path, _ENGINE_KWARGS_DIRECTORY)
+    serialize_engine_kwargs_to_json(engine_kwargs, engine_kwargs_path)
```
Use model_config for this :)

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    from llama_index.core import Settings
+
+    settings_path = os.path.join(path, _SETTINGS_DIRECTORY)
+    serialize_settings_to_json(Settings, settings_path)
```
This needs a dev note comment explaining what is happening here. Why we're accessing a singleton and what we're extracting from it.

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    settings_path = os.path.join(path, _SETTINGS_DIRECTORY)
+    serialize_settings_to_json(Settings, settings_path)
+
+    model_data_path = path
```
useless assign?

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    model_data_path = path
+    _save_model(index, model_data_path)
+
+    flavor_conf = {"engine_type": engine_type}
```
Do we need validation of engine_type?

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+
+    flavor_conf = {"engine_type": engine_type}
+    if _should_add_pyfunc_to_model(index):
+        if mlflow_model.signature is None:
```
Wasn't this done above?

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    )
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        llama_index_version=llama_index.core.__version__,
```
Let's add a utility for this (getting the version)

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+        code=code_dir_subpath,
+        **flavor_conf,
+    )
+    if size := get_total_file_size(path):
```
Validate with the new Resources object type to determine if this is the correct way of doing this now. Parameters that are sent to model serving for validating instance type for serving considerations should conform to the standard interface.

BenWilson2 replied:
See how langchain does Resource definition and copy that for submitting deployment metadata for model serving to use (we will need to make changes in serving for this flavor to consume this metadata obj)

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    index,
+    artifact_path: str,
+    engine_type: str,
+    engine_kwargs: Optional[Dict[str, any]] = None,
```
this should be `model_config`

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+    flavor_conf = _get_flavor_configuration(model_path=local_model_path, flavor_name=FLAVOR_NAME)
+
+    settings_path = os.path.join(local_model_path, _SETTINGS_DIRECTORY)
+    # Settings is a singleton, so it does not need to be passed
```
```suggestion
    # Settings is a singleton
```

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+
+def _load_pyfunc(path):
+    engine_kwargs_path = os.path.join(path, _ENGINE_KWARGS_DIRECTORY)
+    engine_kwargs = deserialize_json_to_engine_kwargs(engine_kwargs_path)
```
read from model_config if defined

BenWilson2 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,385 @@
+
+    flavor_conf = _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME)
+    engine_type = flavor_conf["engine_type"]
+    return _LlamaIndexModelWrapper(_load_model(path, flavor_conf), engine_type, engine_kwargs)
```
extract engine_type and engine_kwargs from model_config by popping. What other args are available for instantiating that we might use model_config kwargs entries for? How will these be handled?

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+_SUPPORTED_SERIALIZATION_METHODS = ["to_dict", "dict"]
+
+
+def _extract_constructor_from_object(o: object) -> Callable:
```
if callable(obj) will replace most of this logic :)

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+        )
+
+
+def _get_dict_method_if_exists(o: object) -> Dict[str, any]:
```
Let's change this to isinstance(BaseClass) for the types that we are certain can be serialized :)

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    raise AttributeError(f"Object {o} does not have a supported deserialization method.")
+
+
+def _get_kwargs(c: Callable) -> Tuple[Set]:
```
Let's verify what LlamaIndex does here with unexpected args. If it throws, we don't need this. If it creates an impossible to diagnose error, we'll need some validation. But for now, we don't need this validation.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+
+def _sanitize_api_key(d: Dict[str, str]) -> Dict[str, str]:
+    # TODO: drop API key and make a note that it must be in an enviornment variable
+    return d
```
after updating this, make sure not to use shorthand variable names like this.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    o_callable = _extract_constructor_from_object(o)
+    required_kwargs, optional_kwargs = _get_kwargs(o_callable)
+    missing_kwargs = set(required_kwargs) - set(o_state_as_dict.keys())
+    if len(missing_kwargs) > 0:
```
Python does this for you.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+        return callable(**kwargs)
+
+
+def _deserialize_json_to_dict_of_objects(path: str) -> Dict[str, any]:
```
rename this to something that is intuitive.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+        json.dump(to_serialize, f, indent=2)
+
+
+def serialize_settings_to_json(settings: Settings, path: str) -> None:
+    _serialize_dict_of_objects_to_json(settings.__dict__, path)
```
Why would we not just call this directly where this is being used?

BenWilson2 replied:
```suggestion
def serialize_settings_to_json(path: str) -> None:
    from llama_index.core import Settings

    _serialized_dict_of_objects_to_json(Settings.__dict__, path)
```

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    for k, v in settings_dict.items():
+        setattr(Settings, k, v)
+
+    return Settings
```
singleton. It's global.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    _serialize_dict_of_objects_to_json(settings.__dict__, path)
+
+
+def deserialize_json_to_settings(path: str) -> Settings:
```
Remove return type

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    return Settings
+
+
+def serialize_engine_kwargs_to_json(engine_kwargs: Dict[str, object], path: str) -> None:
```
This handled by `model_config` serialization.

BenWilson2 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,224 @@
+    _serialize_dict_of_objects_to_json(engine_kwargs, path)
+
+
+def deserialize_json_to_engine_kwargs(path: str) -> Dict[str, object]:
```
same as above.

BenWilson2 commented on mlflow/llama_index/signature.py:
```diff
@@ -0,0 +1,46 @@
+    # out from __init__.py to avoid this.
+    from mlflow.llama_index import _LlamaIndexModelWrapper
+
+    return _LlamaIndexModelWrapper(
```
However, a utility function within init that constructs this can avoid this import. Build and instance of _LlamaIndexModelWrapper via a private func within the init.py

BenWilson2 commented on mlflow/llama_index/signature.py:
```diff
@@ -0,0 +1,46 @@
+def infer_signature_from_input_example(
+    index, example=None, model_config=None, flavor_config=None
+) -> ModelSignature:
+    """
```
We'll need to support the standard OpenAI specific input / output structures ChatCompletions, etc... for a unified llm interface

Commit:
10: PR comments resolved

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
Revert version bump

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
10: simpifly tests and clean up

Signed-off-by: Michael Berk <michael.berk@databricks.com>

Commit:
20: Initial pyfunc logic

Signed-off-by: Michael Berk <michael.berk@databricks.com>

20: Query and retreiver logic. Part of chat logic

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

20: Query and retriever logic - chat not complete

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+        if len(prediction_list) == 1:
```
nit: this holds input not prediction result so might be a bit mislearing

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+                # Assume that any kwarg value that's a sequence of strings with a colon
+                # should be converted to a list of ChatMessage objects.
+                return {k: self._safe_convert_to_chat_messages(v) for k, v in data.items()}
+        elif isinstance(data, (list, np.ndarray)):
+            prediction_list = [self._format_predict_input_chat(d) for d in data]
+        elif isinstance(data, pd.DataFrame):
```
Can we check what are the possible output type from `_convert_llm_input_data`? From a quick read, seems `np.array` and `pd.DataFrame` are not returned.

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+        self.engine_type = engine_type
+        self.predict_callable = self._build_engine_method()
+
+    def _format_predict_input_query_and_retriever(self, data):
```
Is it possible to make this function only accept single data and outputs `QueryBundle` only? Since the prediction is made one by one, I think it's fine to do formatting one by one too, and I believe it makes this function simpler?

This is introspection from LangChain experience, having multiple possible data types (e.g. batch vs single) flying within the data flow makes it super hard to reason about the logic and test with high coverage.

michael-berk replied:
Really great point!

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+            elif len(data) > 1:
+                # Assume that any kwarg value that's a sequence of strings with a colon
+                # should be converted to a list of ChatMessage objects.
+                return {k: self._safe_convert_to_chat_messages(v) for k, v in data.items()}
```
What is the corresponding LlamaIndex invocation signature for this case? What will be the key for each chat message?

michael-berk replied:
There are several signatures, both in terms of number of parameters and the names of the parameters. There is also often a `**kwargs` in the signature. Some examples:
* `def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:`
* `def chat(self, message: str, chat_history: Optional[List[ChatMessage]] = None) -> AgentChatResponse:`
* `def chat(self, query: str, chat_history: Optional[List[ChatMessage]] = None):`

[Search term in llama_index](https://github.com/search?q=repo%3Arun-llama%2Fllama_index+%22def+chat%22+-path%3Alegacy&type=code)

Here's the thought process
* All signatures accept 1 positional argument, so if the dict has `len == 1`, it should be treated as positional
* All signatures require a message history formatted as `List[ChatMessage]`. The user must pass a list of string `[str(ChatMessage(role='user', content='xyz'))]` (llama_index implements a `__repr__` in `ChatMessage`). Then, we parse this via `_safe_convert_to_chat_messages`.

Would appreciate any suggestions on how to make this more robust - it's a start but is not very dynamic 🙂

B-Step62 replied:
Thanks for the explanation, @michael-berk !!

> All signatures accept 1 positional argument, so if the dict has len == 1, it should be treated as positional

Hmm, the `message: str` and `query: str` is not the stringified chat message, but a single question? So the input will look like:

```
model.predict({
    "message": "What is MLflow?"
    "chat_history": [
        {
            "role": "system",
            "content": "xxx",
        },
        {
            "role": "user",
            "content": "yyy",
        }
    ]
```

Also here `chat_history` is passed as list of dictionary, not a list of `ChatMessage`. We need to support both, because in model serving, the request payload is always JSON serialized.

> The user must pass a list of string [str(ChatMessage(role='user', content='xyz'))]

I don't think we can ask that favor to users.... the same input example has to work for  models loaded by `mlflow.llamaindex.load_model()` and `mlflow.pyfunc.load_model()`. Also `str` conversion is not reliable in general, like they (llamaindex) can change `__repr__` implementation to whatever convenient, e.g. omits some field.

My suggestion here is to process the `query`/`message` and other messages separately. What do you think??🤔

```
   messages = []
   if user_query := data.get("query") or data.get("message"):
        messages.append(ChatMessage(role="user", content=user_query)
   if history := data.get("chat_history"):
       messages.extend([ChatMessage(**msg) for msg in history])
```

michael-berk replied:
Nice! Resolved in [this commit](https://github.com/mlflow/mlflow/pull/11674/commits/99478107bb4a3c488794d8139223c20718112d17)

I've added all kwargs that currently exist in llama_index for both `query` and `message_history`.

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+    def _format_predict_input_chat(self, data):
+        """
+        Chat engines can have a variety of signatures. The two primary standards are:
+        1. chat(str, Sequence[ChatMessage])
```
Q: in this case, the former is the question and the latter is the chat history?

michael-berk replied:
Correct

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
 
 
 class _LlamaIndexModelWrapper:
```
Random idea: I found almost all lines in class have conditioned by engine type. What about having subclasses like `_LlamaIndexChatEngineWrapper`?

michael-berk replied:
Genius

michael-berk committed:
20: full pyfunc inference and tests

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
20: pyfunc inference full functionality after comments

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
20: resolve str

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

B-Step62 commented on mlflow/llama_index/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    if not model_config:
+        model_config = {"engine_type": engine_type}
+    elif isinstance(model_config, dict):
+        model_config.update({"engine_type": engine_type})
```
Shall we raise an exception when `model_config` is not dictionary?

B-Step62 commented on mlflow/llama_index/pyfunc_wrapper.py:
```diff
@@ -0,0 +1,159 @@
+
+        if isinstance(input, dict):
+            return self.predict_callable(**input, **(params or {}))
+        else:
```
Can we add explicit `elif` for the other supported type, and raise for unexpected inputs? (or is it complex - like depending on engine type?)

michael-berk replied:
Unfortunately the signature is not consistent, even within a single engine type. I don't see a scalable way to handle unexpected inputs (without essentially recreating python base errors).

B-Step62 commented on mlflow/llama_index/pyfunc_wrapper.py:
```diff
@@ -0,0 +1,159 @@
+
+    @staticmethod
+    def _parse_dict_as_chat_message_objects(data: dict) -> dict:
+        message_matches = (name for name in _CHAT_MESSAGE_PARAMETER_NAMES if name in data)
+        if (key := next(message_matches, None)) and data.get(key):
```
This part is a bit hard to follow, can we simply construct a for loop?
```
for key in CHAT_MESSAGE_PARAMETER_NAMES:
    if v := data.get(key) and not isinstance(v, str):
```

michael-berk replied:
No idea why this is written like this 🙃 

Thanks

B-Step62 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,183 @@
+        if object_as_dict.pop(k, None):
+            _logger.info(
+                "API key removed from object serialization. At inference time,"
+                " the key must be passed as an environment variable or via inference"
+                " parameters."
```
nit: I'd remove `or via inference parameters`.... it will highly likely create a security hole e.g. put API key in the code😅

B-Step62 commented on mlflow/llama_index/serialize_objects.py:
```diff
@@ -0,0 +1,183 @@
+import importlib
```
Shall we split this out to a separate PR?

michael-berk committed:
Half-complete patches of openai

Signed-off-by: Michael Berk <michaelberk99@gmail.com>
